File: MonotoneStack/Leetcode503.java
import java.util.Stack;

public class Leetcode503 {
    public int[] nextGreaterElements(int[] nums){
        int n = nums.length;
        int[] res = new int[n];
        Arrays.fill(res,-1);
        Stack<Integer> stack = new Stack<>();
        for(int i = 0;i <n*2;i++){
            int num = nums[i%n];
            while(!stack.isEmpty() && num > nums[stack.peek()]){
                res[stack.pop()] = num;
            }
            if(i < n)stack.push(i);
        }
        return res;
    }
}

File: MonotoneStack/Python/Leetcode503.py
class Leetcode503:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        res = [-1 for i in range(len(nums))]
        stack = []
        for i in range(2 * len(nums)):
            idx = i % len(nums)
            while stack and nums[stack[-1]] < nums[idx]:
                j = stack.pop()
                res[j] = nums[idx]
            stack.append(idx)
        return res
